### PR TITLE
Wip/finetuning

### PR DESCRIPTION
**Pull request by Mehrad0711**
This PR introduces the following changes:

- Add fine-tuning support for Marian and MBart models.
- Includes more noising functions used in Bart and Mbart pretraining for masked paraphrasing task.
- Temporary fix in tokenization for off-by-one error in Mbart generation. There is no unique solution to this yet (https://github.com/huggingface/transformers/issues/5755)
- Introduces a method to finetune on large datasets (potentially as big as disk memory) by splitting the input file into chunks, loading them to memory one-at-a-time, and training sequentially.

### CODE REVIEW

**Review by gcampax — COMMENTED** [review]
This looks okay to me, but I would want @s-jse to take a look as well. And I'd love to get rid of two copies of the training code, one for "train" and one for "run-lm-finetuning". It's a mess currently.

**Comment by Mehrad0711** [comment]
Same here. The next PRs (transformer bumps) cleanup part of this code. Also now that BART is working in the main train loop, we can start merging the two copies.

File: genienlp/paraphrase/run_lm_finetuning.py [hunk 1]
```diff
@@ -424,7 +462,7 @@ def parse_argv(parser):
     parser.add_argument("--per_gpu_train_batch_size", default=4, type=int,
                         help="Batch size per GPU/CPU for training.")
     parser.add_argument("--per_gpu_eval_batch_size", default=4, type=int,
-                        help="Batch size per GPU/CPU for evaluation.")
+                        help="ze per GPU/CPU for evaluation.")
```
Comment (s-jse) [inline genienlp/paraphrase/run_lm_finetuning.py]: typo?

File: genienlp/paraphrase/run_lm_finetuning.py [hunk 2]
```diff
@@ -435,7 +473,7 @@ def parse_argv(parser):
                         help="Epsilon for Adam optimizer.")
     parser.add_argument("--max_grad_norm", default=1.0, type=float,
                         help="Max gradient norm.")
-    parser.add_argument("--num_train_epochs", default=1.0, type=float,
+    parser.add_argument("--num_train_epochs", default=1, type=int,
```
Comment (s-jse) [inline genienlp/paraphrase/run_lm_finetuning.py:476]: This eliminates the option to train for e.g. 1.5 epochs (I'm not sure if it worked properly before or not), but it is fine since we never used it.

**Review by s-jse — COMMENTED** [review]
Looks good to me too. Just a typo I think.
Yes, we can and should merge training codes. I think we will have to keep generation codes (`predict.py` and `run_generation.py`) separate. We should merge paraphrasing code of run_generation.py into predict.py (it is essentially done) and let `run_generation.py` fully evolve into whatever it has become.

File: genienlp/paraphrase/run_lm_finetuning.py (hunk 1 above)
Comment (Mehrad0711) [reply genienlp/paraphrase/run_lm_finetuning.py]: nice catch.

Comment (Mehrad0711) [reply genienlp/paraphrase/run_lm_finetuning.py]: Sad to bother Travis for one typo.
